### PR TITLE
Allow for multiple simultaneous xfem cuts in an element

### PR DESCRIPTION
**Pull request by austinltam**
ref #15115

Changed the way split works on 2D elements. Split now uses cut plane ids associated to nodes at the element level. Node cut planes ids are assigned based on calls to addEdgeCut, addNodeCut, and addFragmentCut. Also allows for intersections of cutting planes inside elements, creating faceNodes at these intersections.

### CODE REVIEW

**Comment by austinltam** [comment]
@bwspenc Here is my pull request. Thank you for the help!

**Review by bwspenc — COMMENTED** [review]
That non-unity build failure is because you don't have the `#include`s quite right. You can sort it out locally by doing `MOOSE_UNITY=false make`.

**Comment by austinltam** [comment]
@bwspenc I am not sure why the formatting is failing but it doesn't say it makes any changes when I run format-clang.

**Comment by bwspenc** [comment]
@austinltam, since this has been sitting for a while, can you please re-base and push an update to this?

File: modules/xfem/test/tests/solid_mechanics_basic/tests [hunk 1]
```diff
@@ -111,6 +111,15 @@
...
     requirement = 'The XFEM system shall permit branched cracks to be represented in 2D by sequentially cutting a 4-noded quadrilateral element by two prescribed evolving cutting planes'
   [../]
+  [./square_branch_quad_2d_instant]
+    type = Exodiff
+    input = square_branch_quad_2d_instant.i
+    exodiff = 'square_branch_quad_2d_instant_out.e square_branch_quad_2d_instant_out.e-s002'
+    map = false
+    # XFEM requires --enable-unique-ids in libmesh
+    unique_id = true
+    requirement = 'The XFEM system shall permit branched cracks to be represented in 2D by instantly cutting a 4-noded quadrilateral element by two prescribed evolving cutting planes'
```
Comment (bwspenc) [inline modules/xfem/test/tests/solid_mechanics_basic/tests:121]: instantly -> simultaneously

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 2]
```diff
@@ -562,7 +562,7 @@ FancyExtruderGenerator::generate()
     }
   }
 
-  mesh->set_isnt_prepared();
+  mesh->is_prepared();
```
Comment (bwspenc) [inline framework/src/meshgenerators/FancyExtruderGenerator.C:565]: Did you mean to make this change? Seems unrelated.

File: modules/xfem/include/efa/EFAElement2D.h [hunk 3]
```diff
@@ -33,6 +33,10 @@ class EFAElement2D : public EFAElement
   std::vector<std::vector<EFAElement2D *>> _edge_neighbors;
   std::vector<EFAFragment2D *> _fragments;
   std::vector<EFAPoint> _local_node_coor;
+  unsigned int _current_cut_plane_idx = 0;
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAElement2D.h:36]: I know a lot of our code isn't documented like this, but as we add stuff, our convention is to add short doxygen comments for each variable in the header file. I realize this is kind of stating the obvious in this case, but something like this:
```
/// Index of the current cut plane
unsigned int _current_cut_plane_idx = 0;

File: modules/xfem/include/efa/EFAElement2D.h [hunk 4]
```diff
@@ -127,6 +131,26 @@ class EFAElement2D : public EFAElement
 
   std::vector<EFANode *> getCommonNodes(const EFAElement2D * other_elem) const;
 
+  void addInteriorNode(EFAFaceNode * faceNode);
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAElement2D.h:134]: Doxygen comments here too

File: modules/xfem/include/efa/EFAElement2D.h (hunk 3 above)
Comment (bwspenc) [reply modules/xfem/include/efa/EFAElement2D.h:36]: Also, delete the commented out stuff that you didn't end up using both here and below.

File: modules/xfem/include/efa/EFAElement2D.h [hunk 5]
```diff
@@ -127,6 +131,26 @@ class EFAElement2D : public EFAElement
...
+  bool isInteriorNode(EFANode * node) const;
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAElement2D.h:135]: Functions like this should be documented using the function-style doxygen comments with brief descriptions of what the functions do and what the arguments to them are. Describe the parameters with `@param` and the return (if applicable) with `@return`. Look at the declaration of `addGeomMarkedElem2D` in XFEM.h for an example.

File: modules/xfem/include/efa/EFAElement3D.h [hunk 6]
```diff
@@ -155,6 +155,7 @@ class EFAElement3D : public EFAElement
                           bool add_to_neighbor,
                           bool add_to_adjacent);
   std::vector<EFANode *> getCommonNodes(const EFAElement3D * other_elem) const;
+  unsigned int getNewCutPlaneIdx();
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAElement3D.h:158]: Document this one

File: modules/xfem/include/efa/EFAFragment2D.h [hunk 7]
```diff
@@ -30,6 +30,8 @@ class EFAFragment2D : public EFAFragment
 private:
   EFAElement2D * _host_elem;
   std::vector<EFAEdge *> _boundary_edges;
+  std::vector<unsigned int> _frag_cut_plane_idx;
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAFragment2D.h:33]: Add documentation

File: modules/xfem/src/base/XFEM.C [hunk 8]
```diff
@@ -581,6 +582,7 @@ XFEM::markCutEdgesByState(Real time)
        ++pmeit)
   {
     const Elem * elem = pmeit->first;
+
```
Comment (bwspenc) [inline modules/xfem/src/base/XFEM.C:585]: Remove extra blank line.

File: modules/xfem/src/base/XFEM.C [hunk 9]
```diff
@@ -420,11 +419,13 @@ XFEM::markCutEdgesByGeometry()
                                                 gmei._frag_cut_edges[i]._distance))
           {
             marked_edges = true;
-            if (!isElemAtCrackTip(gme.first))
-              _has_secondary_cut = true;
+            //            if (!isElemAtCrackTip(gme.first))
+            //              _has_secondary_cut = true;
```
Comment (bwspenc) [inline modules/xfem/src/base/XFEM.C:423]: If this is a change we really want to make, just delete the lines.

File: modules/xfem/src/efa/EFAEdge.C [hunk 10]
```diff
@@ -406,3 +406,31 @@ EFAEdge::removeEmbeddedNode(EFANode * node)
...
+  // check _edge_node1 for node
+  if (node == _edge_node1)
+    hasNode = true;
+
+  // check _edge_node2 for node
+  if (node == _edge_node2)
+    hasNode = true;
+
+  //  //check embedded nodes for node
+  //  for (unsigned int inode = 0; inode < _embedded_nodes.size(); ++inode)
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAEdge.C:424]: Are there no embedded nodes on the edge? If you don't need this code, delete it.

File: modules/xfem/src/efa/EFAElement2D.C [hunk 11]
```diff
@@ -466,6 +468,7 @@ EFAElement2D::clearNeighbors()
 void
 EFAElement2D::setupNeighbors(std::map<EFANode *, std::set<EFAElement *>> & InverseConnectivityMap)
 {
+
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAElement2D.C:471]: Remove extra new line

File: modules/xfem/src/efa/EFAElement2D.C [hunk 12]
```diff
@@ -593,6 +596,7 @@ EFAElement2D::initCrackTip(std::set<EFAElement *> & CrackTipElements)
 unsigned int
 EFAElement2D::getCrackTipSplitElementID() const
 {
+
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAElement2D.C:599]: remove line

File: modules/xfem/src/efa/EFAElement2D.C [hunk 13]
```diff
@@ -836,8 +840,8 @@ EFAElement2D::updateFragments(const std::set<EFAElement *> & CrackTipElements,
   unsigned int num_cut_frag_edges = _fragments[0]->getNumCuts();
   unsigned int num_cut_nodes = _fragments[0]->getNumCutNodes();
   unsigned int num_frag_edges = _fragments[0]->numEdges();
-  if (num_cut_frag_edges > 3)
-    EFAError("In element ", _id, " there are more than 2 cut fragment edges");
+  //  if (num_cut_frag_edges > 3)
+  //    EFAError("In element ", _id, " there are more than 2 cut fragment edges");
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAElement2D.C:844]: Remove

File: modules/xfem/src/efa/EFAElement2D.C [hunk 14]
```diff
@@ -2067,3 +2099,331 @@ EFAElement2D::getCommonNodes(const EFAElement2D * other_elem) const
...
+  _interior_nodes.push_back(faceNode);
+}
+
+bool
+EFAElement2D::isInteriorNode(EFANode * node) const
+{
+  for (unsigned int i = 0; i < _interior_nodes.size(); ++i)
+  {
+    if (node == _interior_nodes[i]->getNode())
+    {
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAElement2D.C:2115]: Our convention is to not have curly braces around one-liner blocks.

File: modules/xfem/src/efa/EFAElement2D.C [hunk 15]
```diff
@@ -2067,3 +2099,331 @@ EFAElement2D::getCommonNodes(const EFAElement2D * other_elem) const
...
+EFAFaceNode *
+EFAElement2D::getFaceNode(EFANode * node) const
+{
+  EFAFaceNode * faceNode = nullptr;
+  EFANode * interiorNode = nullptr;
+  for (unsigned int i = 0; i < _interior_nodes.size(); ++i)
+  {
+    interiorNode = _interior_nodes[i]->getNode();
+    if (node == interiorNode)
+    {
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAElement2D.C:2131]: Remove braces around one liner.

File: modules/xfem/src/efa/EFAElement2D.C [hunk 16]
```diff
@@ -2067,3 +2099,331 @@ EFAElement2D::getCommonNodes(const EFAElement2D * other_elem) const
...
+        return;
+    }
+    if (_cut_plane_nodes[index].size() >= 2)
+      addNodeToCutPlaneIdx(node, cut + 1);
+    else
+      _cut_plane_nodes[index].push_back(node);
+  }
+}
+
+// void
```
Comment (bwspenc) [inline modules/xfem/src/efa/EFAElement2D.C:2198]: Remove commented code

**Review by bwspenc — COMMENTED** [review]
These comments are on mostly superficial formatting things. I'll need to pull this down and dig deeper into this, but go ahead and start doing these clean ups across your code.

**Comment by bwspenc** [comment]
Rebased and force-pushed to trigger a new build against the updated libmesh that was just put in. Hopefully this will fix that build failure on the set_union.

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Comment by bwspenc** [comment]
Commenting to keep this open. @austinltam I know you have some updates for this. Just a reminder to push those when you get a chance.

**Comment by bwspenc** [comment]
@austinltam I ran clang-format on your commit, fixed XFEM.C to have unix rather than dos end lines, rebased it against the latest devel branch, and pushed that update just now.

File: modules/xfem/include/efa/EFAElement2D.h [hunk 17]
```diff
@@ -127,6 +133,96 @@ class EFAElement2D : public EFAElement
...
+  void addNodeToCutPlaneIdx(EFANode * node, unsigned int cut);
+  /**
+   * Reassign input node to a different cut plane index
+   * @param node   The EFANode to be reassigned
+   * @param unsigned int of cut plane index that node is to be reassigned to
+   */
+  void reassignNodeToCutPlaneIdx(EFANode * node, unsigned int cut);
+  /**
+   * Get all cut plane nodes in a vector of vector
+   * @return vector of vectors containing all cut nodes binned by cut plane index
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAElement2D.h:185]: 'binned'->'ordered'

File: modules/xfem/include/efa/EFAFragment2D.h [hunk 18]
```diff
@@ -51,5 +55,59 @@ class EFAFragment2D : public EFAFragment
...
+   * intersections
+   * @return vector of fragments
+   */
+  std::vector<EFAFragment2D *> split(std::map<unsigned int, EFANode *> & EmbeddedNodes);
+  /**
+   * Splits fragment into many fragments depending on cuts
+   * @param EmbeddedNodes is used to create new intersection nodes in the case of cut plane
+   * intersections
+   * @param vector of fragment cut plane ids
+   * @param vector of vectors with each cut node binned by cut planes
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAFragment2D.h:70]: 'binned'->'ordered'

File: modules/xfem/include/efa/EFAFragment2D.h [hunk 19]
```diff
@@ -51,5 +55,59 @@ class EFAFragment2D : public EFAFragment
...
+  /**
+   * Determine if two lines intersect
+   * @param two length vector with coordinates of point 1
+   * @param two length vector with coordinates of point 2
+   * @param two length vector with coordinates of point 1 of cutting line
+   * @param two length vector with coordinates of point 2 of cutting line
+   * @param two length vector with coordinates of intersection point if it exists
+   * @return bool indicating if lines intersect
+   */
+  bool IntersectSegmentWithCutLine(const std::vector<double> & segment_point1,
```
Comment (bwspenc) [inline modules/xfem/include/efa/EFAFragment2D.h:85]: Start method name with lower case to follow MOOSE style. (`intersectSegmentWithCutLine`)

Comment (bwspenc) [inline modules/xfem/include/efa/EFAFragment2D.h:85]: A few of these methods already exist in XFEMFuncs.C. Unless there's a really good reason to reimplement them here, you should use those, or add missing functionality to them if needed.

**Pull request closed without merging**